### PR TITLE
remove loop as argument and update examples

### PR DESCRIPTION
**Pull request by r0fls**
This entirely removes the `loop` as an argument, from tests, examples, and sanic core. Since it's no longer going to be supported by asyncio (https://github.com/channelcat/sanic/issues/152#issuecomment-261316341) it doesn't make sense to *go against the grain*. All of the example modifications except the peewee one have been tested to work.

We'll probably want to issue a deprecation message for a few weeks before merging this.

### CODE REVIEW

**Comment by seemethere** [comment]
I'm a fan but let's leave this up for at least a week so that people can voice their opinions about it as well. If we don't find a compelling reason not to merge this change in by the end of the week this will be rolled into 0.3.0.

**Comment by r0fls** [comment]
Tagging some of the open issues that are related:
https://github.com/channelcat/sanic/issues/305
https://github.com/channelcat/sanic/issues/152

**Comment by youknowone** [comment]
Please see #338 also.

**Comment by ubergarm** [comment]
It's definitely cleaner and less confusing to explicitly specify the loop everywhere. So :+1: 

It would be good to have another similar example showing how to limit the number of concurrent coroutines running using Semaphore or however is best to do so.

**Comment by r0fls** [comment]
@youknowone that should be addressed by https://github.com/channelcat/sanic/issues/275. I have a half finished PR for that.

**Comment by r0fls** [comment]
@ubergarm your implementation looks like the best idea we have for that, as of now. I'll add it as an example.

File: examples/sanic_aiopg_sqlalchemy_example.py [hunk 1]
```diff
@@ -34,40 +30,37 @@
...
 
 
 @app.route("/")
 async def handle(request):
-    async with engine.acquire() as conn:
-        result = []
-        async for row in conn.execute(polls.select()):
-            result.append({"question": row.question, "pub_date": row.pub_date})
-        return json({"polls": result})
+    async with create_engine(connection) as engine:
```
Comment (Tim-Erwin) [inline examples/sanic_aiopg_sqlalchemy_example.py]: Where is `create_engine` defined?
Do we really want to recommend to create a new engine in each request?

**Comment by Tim-Erwin** [comment]
While in general this seems to be a very good idea, it breaks applications where libraries need to use the same loop Sanic is running in. Let's assume I'm using an async database client which gets initialized before `run`ning the app, how is it supposed to get hold of the loop which is only created later? Do we need to use `before_start` for that? If so, this should at least be documented somehow.

File: examples/sanic_asyncpg_example.py [hunk 2]
```diff
@@ -21,45 +19,40 @@
...
+                        (id, content, post_date) VALUES ({i}, {i}, now())""")
 
 
 @app.route("/")
 async def handler(request):
-    async with pool.acquire() as connection:
-        async with connection.transaction():
-            results = await connection.fetch('SELECT * FROM sanic_post')
-            return json({'posts': jsonify(results)})
+    async with create_pool(**DB_CONFIG) as pool:
```
Comment (r0fls) [inline examples/sanic_asyncpg_example.py]: "Also I think create_pool is not defined." see line 8

File: examples/sanic_aiopg_sqlalchemy_example.py (hunk 1 above)
Comment (r0fls) [inline examples/sanic_aiopg_sqlalchemy_example.py]: "Where is create_engine defined?" See line 9

Comment (Tim-Erwin) [inline examples/sanic_aiopg_sqlalchemy_example.py]: Ah, sorry, missed the top part of the file

**Comment by r0fls** [comment]
@Tim-Erwin thanks for your review and feedback. Regarding:

>it breaks applications where libraries need to use the same loop Sanic is running in

One possibility in that scenario is using this new feature:
https://github.com/channelcat/sanic/pull/348

File: examples/sanic_aiopg_sqlalchemy_example.py (hunk 1 above)
Comment (r0fls) [inline examples/sanic_aiopg_sqlalchemy_example.py]: I get caught by that all of the time, but that style of import is the way throughout sanic.

File: sanic/sanic.py [hunk 3]
```diff
@@ -275,8 +274,8 @@ def converted_response_type(self, response):
 
     def run(self, host="127.0.0.1", port=8000, debug=False, before_start=None,
             after_start=None, before_stop=None, after_stop=None, ssl=None,
-            sock=None, workers=1, loop=None, protocol=HttpProtocol,
-            backlog=100, stop_event=None, register_sys_signals=True):
+            sock=None, workers=1, protocol=HttpProtocol, backlog=100,
```
Comment (seemethere) [inline sanic/sanic.py]: Do you think it would be smart to leave the `loop` param (with a deprecation notice from `warnings.warn`) in for `0.3.0` and then officially remove it in the next release?

Comment (r0fls) [inline sanic/sanic.py]: @seemethere https://github.com/channelcat/sanic/pull/350

File: sanic/sanic.py [hunk 4]
```diff
@@ -296,18 +295,11 @@ def run(self, host="127.0.0.1", port=8000, debug=False, before_start=None,
         :param sock: Socket for the server to accept connections from
         :param workers: Number of processes
                         received before it is respected
-        :param loop: asyncio compatible event loop
         :param protocol: Subclass of asyncio protocol class
         :return: Nothing
         """
         self.error_handler.debug = debug
         self.debug = debug
-        if loop is not None:
```
Comment (seemethere) [inline sanic/sanic.py]: I think keeping them in for `0.3.0` and then taking them off for the next big version change will be the better choice here. Keep the messages but not the functionality of the loop is what I meant.

Comment (r0fls) [inline sanic/sanic.py]: aaah, I see 👍  that is an even better idea, since I want to see the loop removed as soon as possible -- but wait that also doesn't make sense: if it's already been deprecated, why give a deprecation warning? 🤔

Comment (r0fls) [inline sanic/sanic.py]: I'm confused 🤔  :octocat:

Comment (r0fls) [inline sanic/sanic.py]: Oh, I think I get it -- update this PR to leave the loop as an argument, but don't do anything and also print the message: is that what you have in mind? I can get behind that 😄

Comment (seemethere) [inline sanic/sanic.py]: Basically, with this PR we remove the loop functionality. But we keep the `loop` param but it will do nothing except trigger the warnings message so that anyone who was using the `loop` param won't just get a `TypeError` Exception and not know why. Then at the next big release say `0.4.0` we remove the param since we gave people enough of a heads up.

A lot of people just install sanic with `pip install sanic` so not having enough warning can potentially kill a lot of the apps people have built.

TL;DR

```python
if loop is not None:
    warnings.warn((
        '`loop` parameter will be removed after '
        'v`0.3.0` make sure to remove it before the next release!'))
```

**Comment by seemethere** [comment]
Alright without a compelling argument against we will roll this into `0.3.0` scheduled for release tomorrow!

**Comment by Tim-Erwin** [comment]
Wonderful. However, as mentioned in the before, it doesn't seem to be a good idea to create new connection pools in each request. Hence, we should not suggest to do that in our examples. While examples are just that - examples - they should give a *good* starting point. I think many will struggle with this part.

**Comment by r0fls** [comment]
@Tim-Erwin I'm curious to see some performance statistics to back the claim that this is poor practice. While I agree, if you can avoid opening a connection for each query, that is better. However, It seems that passing the loop is not really supported by Python, so we *need* to move away from that. It's less of a performance concern than a sheer "this doesn't work" situation. You're welcome to create a PR or provide a suggestion to improve the examples. They're not going to be set in stone after this.

**Comment by Tim-Erwin** [comment]
@r0fls The reason to use a connection pool is to avoid creating new connections over and over again, right? So even without any difference in performance creating a pool for just one request contradicts its purpose. I understand that passing the loop is no longer supported and that's a good thing, because IMHO the loop should be invisible as long as you don't intend to tinker with it.

In my current project with Sanic I'm explicitly creating the loop and passing it in because my DB client (which I create before running the app) has to use the same loop. So I'm really not an expert in how to do it differently. If I understand correctly we'd have to pass in a function `before_start` that's then provided the loop which we can use to init other libraries?

**Comment by go2sh** [comment]
You have to remember that every context (process, thread) etc. has its own loop. You cannot share a loop over thread or process borders. So you need create a connection pool for every worker thread or process you have. From this point of view, the approach from @r0fls is correct.

But now the `before_start` function kicks in. Here you can create a connection per worker since it is called in the context of the worker (process). It shouldn't event need the loop argument as you can use `get_event_loop`.

But you are right, for a single request a connection pool is overkill since only one connection can be used.
